src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,401 @@
+#include "ExtractTileOperations.h"
+
+#include "IRMatch.h"  // expr_match
+#include "IRMutator.h"
+#include "IROperator.h"  // Expr + Expr
+#include "Util.h"        // ScopedValue
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::vector;
+
+namespace {
+
+template<int Dim>
+struct Tile {
+    bool result;
+    Expr base;
+    Expr stride[Dim];
+    int extent[Dim];
+};
+
+enum class AMXOpType {
+    Int8,
+    Bf16,
+};
+
+const auto wild_i32 = Variable::make(Int(32), "*");
+const auto wild_i32x = Variable::make(Int(32, 0), "*");
+
+Tile<2> is_2d_tile_index(const Expr &e) {
+    // ramp(ramp(base, 1, 4), x4(stride), 4)
+    vector<Expr> matches;
+    if (const auto *r1 = e.as<Ramp>()) {
+        if (const auto *r2 = r1->base.as<Ramp>()) {
+            auto ramp_2d_pattern = Ramp::make(Ramp::make(wild_i32, wild_i32, r2->lanes), Broadcast::make(wild_i32, r2->lanes), r1->lanes);
+            if (expr_match(ramp_2d_pattern, e, matches)) {
+                return {true, std::move(matches[0]), {std::move(matches[2]), std::move(matches[1])}, {r1->lanes, r2->lanes}};
+            }
+        }
+    }
+    return {};
+}
+
+Tile<3> is_3d_tile_index(const Expr &e) {
+    vector<Expr> matches;
+    auto add_sub_pattern = (wild_i32x + wild_i32x) - wild_i32x;
```

Comment:
The engineer who has implemented above code has left. We are going to investigate it shortly.

Reply:
> I'm surprised this is an add-sub. Can you comment why this is? Is it perhaps overfit to a particular index calculation?

The add-sub pattern above is not introduced as part of support for AMX in this PR. The pattern is related to the way intermediate index is created for accessing elements, when vectorization is in scheduling and the function input is of types like ImageParam<> or Buffer<>. 

The similar pattern can be seen in tutorials/lesson_10_aot_compilation_generate.cpp when built on a host with NO support for AMX.

Reply:
I see, this is probably subtracting the mins. Does the test still work if we set the buffer min to 0? (In the test, add `A.dim(0).set_min(0).dim(1).set_min(0); B.dim(0).set_min(0).dim(1).set_min(0)`)

Reply:
It hit the asset at line 362 of the same file in function  *Stmt visit(const Store *op) override*
**Store to AMX tile allocation of a non-tile value**

Reply:
I think this is likely because the index expression doesn't match the pattern being looked for any more (a + b - c), because c is 0. I think we need the test to cover this case (in addition to a case where the mins vary), and fix it so it works.

I think maybe using patterns for looking for the right index expressions is not going to work very well. Maybe what we need to do instead is write some logic to dig through the needed expressions, with more tolerance for variations?

Reply:
concerning this I was wondering if it would be a good idea to use the `simplify` operation and then perform the pattern matching on the resulting expression?

Reply:
I've adjusted the index expression checker to be more robust against those 2 scenarios